crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+            let mut new_map = pending_blocks.clone();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        }
+        self.update_internal_state();
+    }
+
+    // remove all data in the queue below a given height. this does not remove
+    // data from the soft queue, only the pending queue, and updates the head height.
+    // TODO: add error handling
+    fn remove_data_below_height(&mut self, height: Height) {
+        // remove all data below the new incoming block from the pending data
+        let tmp_pending = self.pending_blocks.clone();
+        let mut pending_keys: Vec<&Height> = tmp_pending.keys().collect();
+        pending_keys.sort();
+        for key in pending_keys {
+            if *key < height {
+                self.pending_blocks.remove(key);
+            }
+        }
+    }
+
+    /// This function organizes the internal state of the queue based on the
+    /// tendermint/CometBTF fork choice rules.
+    ///
+    /// Once a block is added to the pending_blocks in the queue, this function
+    /// is called. It walks the pending blocks from lowest to highest height,
+    /// checking to see if there is a continuous chain of blocks. For every block
+    /// that is a descendant of the most recent "soft" block, and has a direct
+    /// descendant, that block gets added to the `soft_blocks` BTreeMap and the
+    /// head height is updated.
+    fn update_internal_state(&mut self) {
+        // check if the block added connects blocks in the pending queue
+        let mut heights: Vec<Height> = self.pending_blocks.keys().cloned().collect();
+        heights.sort();
+        // walk the pending blocks starting from the head height
+        for height in heights {
+            // if the very first height in the pending blocks is the head height
+            if height == self.head_height {
+                if let Some(pending_blocks) = self.pending_blocks.clone().get(&height) {
+                    // walk the pending blocks at that height and check if any of them are a parent
+                    for block in pending_blocks.values() {
+                        if self.is_block_a_parent(block.clone()) {
+                            self.soft_blocks.insert(height, block.clone());
+                            self.most_recent_soft_hash = block.block_hash();
+                            self.head_height = height.increment();
+                            self.remove_data_below_height(self.head_height);
+                            break;
+                        }
+                    }
+                }
+            // if the first height in the queue is not the head height just stop reorg
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Can you write tests for this queue in isolation to demonstrate usage and edge cases? The tests can be smaller and more focused and don't need a mocked executor.

Reply:
yes i would like this, can you test for cases such as:
1. adding a block 1 height above the committed height but the parent hash is not the committed block (should not get added)
2. adding a block N+2 where the committed height is N, then committing N+1 which is not a parent of N+2 (N+2 should be removed)
3. committing a block at some height should remove all other blocks at that height and below

by "committing" above i mean finalizing, but i'm not sure if that corresponds to "head/soft" in this queue or not